sympy/core/function.py

```diff
@@ -855,57 +860,32 @@ class Derivative(Expr):
 
     Derivative wrt non-Symbols:
 
-    This class also allows derivatives wrt non-Symbols that have _diff_wrt
+    This class don't allows derivatives wrt non-Symbols that have _diff_wrt
     set to True, such as Function and Derivative. When a derivative wrt a non-
     Symbol is attempted, the non-Symbol is temporarily converted to a Symbol
     while the differentiation is performed.
 
-    Note that this may seem strange, that Derivative allows things like
-    f(g(x)).diff(g(x)), or even f(cos(x)).diff(cos(x)).  The motivation for
-    allowing this syntax is to make it easier to work with variational calculus
-    (i.e., the Euler-Lagrange method).  The best way to understand this is that
-    the action of derivative with respect to a non-Symbol is defined by the
-    above description:  the object is substituted for a Symbol and the
-    derivative is taken with respect to that.  This action is only allowed for
-    objects for which this can be done unambiguously, for example Function and
-    Derivative objects.  Note that this leads to what may appear to be
-    mathematically inconsistent results.  For example::
-
         >>> from sympy import cos, sin, sqrt
         >>> from sympy.abc import x
         >>> (2*cos(x)).diff(cos(x))
-        2
-        >>> (2*sqrt(1 - sin(x)**2)).diff(cos(x))
-        0
+        Traceback (most recent call last):
+        ...
+        ValueError: Can't calculate 1st derivative wrt cos(x).
 
-    This appears wrong because in fact 2*cos(x) and 2*sqrt(1 - sin(x)**2) are
-    identically equal.  However this is the wrong way to think of this.  Think
+    However this is the wrong way to think of this.  Think
```

Comment:
This whole paragraph needs to be rewritten. The "inconsistency" mostly goes away with undefined functions because they don't have any assumed identities on them, so it just needs to be clear that this works simply by replacing the function by a symbol and differentiating wrt that.